RoddyCore/src/de/dkfz/roddy/tools/ComplexLine.groovy

```diff
@@ -0,0 +1,196 @@
+            List<StringBuilder> splittedBlocks = []
+            for (int i = 0; i < blocks.size(); i++) {
+                List<String> splitted = null
+                if (blocks[i] instanceof TextNode) {
+                    splitted = blocks[i].splitBy(s) as List<String>
+
+                    // If our string ends with the splitter, the splitted entry would be lost. Append an empty entry to preserve it.
+                    if (blocks[i].reassemble().endsWith(s))
+                        splitted = splitted + [""]
+                }
+                if (blocks[i] instanceof ComplexNode) {
+                    splitted = [blocks[i].reassemble()] // Complex nodes will just be reassembled. We don't want to split over multi levels!
+                }
+
+                // Convert the found strings to builders
+                List<StringBuilder> newBuilders = splitted.collect { new StringBuilder(it) }
+
+                // If there already is a splitted entry, append the first new found string to the last entry.
+                if (splittedBlocks) {
+                    splittedBlocks[-1] << newBuilders[0]
+                    if (newBuilders.size() > 1)
+                        splittedBlocks += newBuilders[1..-1]
+                } else
+                    splittedBlocks += newBuilders // If there is nothing, just take all entries as new entries.
+            }
+            // Collect, convert, return
+            def res = splittedBlocks.collect { new String(it) }
+            return res as String[]
+        }
+    }
+
+
+    public static isOpeningOrClosingCharacter(Character c) {
+        return possibleContainersByOpeningCharacter.containsKey(c) || possibleContainersByClosingCharacter.containsKey(c)
+    }
+
+    public static fitsToOpeningCharacter(Character c, Character o) {
+        return possibleContainersByClosingCharacter[c] == o
+    }
+
+    public static LineNode parseLine(String line) {
+
+        Stack<Character> openedStarts = new Stack<>();
+        openedStarts.push('ö' as char)
+
+        ComplexNode current = new ComplexNode();
+        Stack<ComplexNode> lineNodes = new Stack<>()
+        lineNodes.push(current)
+
+        for (int i = 0; i < line.length(); i++) {
+
+            char character = line.getChars()[i];
+
+            if (!isOpeningOrClosingCharacter(character)) {
+                ((TextNode) current.blocks[-1]).text << character;
+                continue;
+            } else if (fitsToOpeningCharacter(character, openedStarts.peek())) {
+                openedStarts.pop()
+                if (lineNodes.size() > 0)
+                    lineNodes.pop()
+                if (lineNodes.size() > 0)
+                    current = lineNodes.peek()
+                current.blocks << new TextNode(character as String);
+            } else {
+                ((TextNode) current.blocks[-1]).text << character;
+                openedStarts.push(character)
```

Comment:
There is no check here that this character is indeed an opening character. It could also be closing, namely if it is neither a normal character nor the closing character matching to the top of the openedStarts stack.

What would happen with the input 'blablabla)'?  (oops forgot the opening paren!) Should there not be at least a warning in this situation?